File: server/app/cefr.py
"""Server-side heuristic CEFR (A1–C2) estimator for a single sentence.

Same three axes and weights as the browser's src/cefr.js — vocabulary rarity
(frequency lists), sentence length, syntactic complexity — but the syntax axis
uses a real dependency parse (subordinate-clause count) instead of counting
commas, and it's symmetric across PT and DE. Output shape matches the frontend
exactly: {level, index, factors, metrics:{words, vocab}}, with `factors` as the
same Portuguese strings, so the tooltip reads identically regardless of engine.
"""
import math
import re

from .freq import rank_map, band_from_rank
from .nlp import get_nlp

LEVELS = ["A1", "A2", "B1", "B2", "C1", "C2"]

# Unicode letters, allowing internal apostrophe/hyphen. re matches accented
# letters via [^\W\d_].
_TOKEN_RE = re.compile(r"[^\W\d_][^\W\d_'\-]*", re.UNICODE)

# Dependency labels that head a subordinate/embedded clause (UD-style, as used
# by the pt/de core_news models).
_SUBORD_DEPS = {
    "advcl", "acl", "relcl", "acl:relcl", "ccomp", "xcomp", "csubj", "csubjpass",
}

# Fallback single-word subordinators (only used when no spaCy model is present).
_SUBORDINATORS = {
    "de": {"dass", "weil", "wenn", "als", "ob", "obwohl", "während", "damit",
           "bevor", "nachdem", "sobald", "solange", "seitdem", "falls", "indem",
           "sodass", "obgleich", "sofern", "wohingegen"},
    "pt": {"que", "porque", "quando", "se", "embora", "enquanto", "como",
           "conforme", "caso", "pois", "contudo", "todavia", "porquanto",
           "conquanto", "porém", "senão", "consoante"},
}


def _round_half_up(x: float) -> int:
    return math.floor(x + 0.5)


def _features(text: str, lang: str):
    """Return (words, subordinate_clause_count). Uses spaCy when available."""
    nlp = get_nlp(lang)
    if nlp is not None:
        doc = nlp(text)
        words = [t.text.lower() for t in doc if t.is_alpha]
        subs = sum(1 for t in doc if t.dep_ in _SUBORD_DEPS)
        return words, subs
    # Fallback: regex tokens + subordinator/comma heuristic (like the frontend).
    words = _TOKEN_RE.findall(text.lower())
    subordinators = _SUBORDINATORS.get(lang, set())
    subs = sum(1 for w in words if w in subordinators) + text.count(",")
    return words, subs


def assess_sentence(text: str, lang: str):
    words, subs = _features(text, lang)
    n = len(words)
    if n == 0:
        return None

    ranks = rank_map(lang)
    bands = [band_from_rank(ranks.get(w)) for w in words]

    # Vocabulary: mean difficulty of the hardest quartile (min 1 word), so one
    # rare word doesn't slam the whole sentence to C2.
    hard_count = max(1, math.ceil(n * 0.25))
    vocab = sum(sorted(bands, reverse=True)[:hard_count]) / hard_count

    # Length axis (word count).
    length = (1 if n <= 6 else 2 if n <= 10 else 3 if n <= 15
              else 4 if n <= 22 else 5 if n <= 30 else 6)

    # Syntax axis from subordinate-clause count.
    syntax = (1 if subs == 0 else 3 if subs == 1 else 4 if subs == 2
              else 5 if subs == 3 else 6)

    composite = 0.55 * vocab + 0.30 * length + 0.15 * syntax
    index = min(5, max(0, _round_half_up(composite) - 1))
    level = LEVELS[index]

    # Hardest words (band >= 4) for the tooltip — most difficult first, deduped.
    hard_words = []
    seen = set()
    for w, b in sorted(zip(words, bands), key=lambda x: x[1], reverse=True):
        if b >= 4 and w not in seen:
            seen.add(w)
            hard_words.append(w)
        if len(hard_words) == 4:
            break

    factors = []
    if vocab >= 4:
        factors.append(
            f"Vocabulário avançado: {', '.join(hard_words)}"
            if hard_words else "Vocabulário pouco frequente"
        )
    elif vocab >= 3:
        factors.append("Vocabulário intermediário")
    else:
        factors.append("Vocabulário comum")
    if length >= 4:
        factors.append(f"Sentença longa ({n} palavras)")
    if subs >= 1:
        factors.append(f"{subs} oração(ões) subordinada(s)")

    return {
        "level": level,
        "index": index,
        "factors": factors,
        "metrics": {"words": n, "vocab": round(vocab, 1)},
    }

File: src/cefr.js
// Heuristic CEFR (A1–C2) estimator for a single sentence.
// Combines three axes: vocabulary rarity (from word-frequency lists — the most
// predictive), sentence length, and syntactic complexity (clause markers).
// This is an approximation, as all sentence-level CEFR scoring inherently is.
import freqDe from './data/freq-de.js';
import freqPt from './data/freq-pt.js';

export const LEVELS = ['A1', 'A2', 'B1', 'B2', 'C1', 'C2'];

const RAW = { de: freqDe, pt: freqPt };
const RANK_MAPS = {}; // lang -> Map(word -> rank index)

function rankMap(lang) {
  if (!RANK_MAPS[lang]) {
    const words = (RAW[lang] ?? '').split(' ');
    const m = new Map();
    for (let i = 0; i < words.length; i++) m.set(words[i], i);
    RANK_MAPS[lang] = m;
  }
  return RANK_MAPS[lang];
}

// Single-word subordinating conjunctions — a rough clause-complexity signal.
const SUBORDINATORS = {
  de: new Set(['dass', 'weil', 'wenn', 'als', 'ob', 'obwohl', 'während', 'damit',
    'bevor', 'nachdem', 'sobald', 'solange', 'seitdem', 'falls', 'indem',
    'sodass', 'obgleich', 'sofern', 'wohingegen']),
  pt: new Set(['que', 'porque', 'quando', 'se', 'embora', 'enquanto', 'como',
    'conforme', 'caso', 'pois', 'contudo', 'todavia', 'porquanto', 'conquanto',
    'porém', 'senão', 'consoante']),
};

// Map a frequency rank to a difficulty band 1 (A1) … 6 (C2).
function bandFromRank(rank) {
  if (rank == null) return 6;      // outside the top list → rare
  if (rank < 500) return 1;
  if (rank < 1000) return 2;
  if (rank < 2000) return 3;
  if (rank < 3500) return 4;
  if (rank < 6000) return 5;
  return 6;
}

function tokenize(text) {
  return text.toLowerCase().match(/[\p{L}][\p{L}''-]*/gu) ?? [];
}

/**
 * @returns {{level, index, factors, metrics} | null}
 */
export function assessSentence(text, lang) {
  const words = tokenize(text);
  const n = words.length;
  if (n === 0) return null;

  const map = rankMap(lang);
  const bands = words.map(w => bandFromRank(map.get(w)));

  // Vocabulary: average difficulty of the hardest quartile (min 1 word),
  // so a single rare word doesn't slam the whole sentence to C2.
  const sortedDesc = [...bands].sort((a, b) => b - a);
  const hardCount = Math.max(1, Math.ceil(n * 0.25));
  const vocab = sortedDesc.slice(0, hardCount).reduce((s, b) => s + b, 0) / hardCount;

  // Length axis (word count).
  const length = n <= 6 ? 1 : n <= 10 ? 2 : n <= 15 ? 3 : n <= 22 ? 4 : n <= 30 ? 5 : 6;

  // Syntax axis: subordinating conjunctions + commas (clause boundaries).
  const subs = words.filter(w => SUBORDINATORS[lang]?.has(w)).length;
  const commas = (text.match(/,/g) ?? []).length;
  const markers = subs + commas;
  const syntax = markers === 0 ? 1 : markers === 1 ? 3 : markers === 2 ? 4 : markers === 3 ? 5 : 6;

  const composite = 0.55 * vocab + 0.30 * length + 0.15 * syntax;
  const index = Math.min(5, Math.max(0, Math.round(composite) - 1));
  const level = LEVELS[index];

  // Hardest words (B2+) for the tooltip — most difficult first, deduped.
  const hardWords = [...new Set(
    words
      .map((w, i) => ({ w, b: bands[i] }))
      .filter(x => x.b >= 4)
      .sort((a, b) => b.b - a.b)
      .map(x => x.w)
  )].slice(0, 4);

  const factors = [];
  if (vocab >= 4) {
    factors.push(hardWords.length
      ? `Vocabulário avançado: ${hardWords.join(', ')}`
      : 'Vocabulário pouco frequente');
  } else if (vocab >= 3) {
    factors.push('Vocabulário intermediário');
  } else {
    factors.push('Vocabulário comum');
  }
  if (length >= 4) factors.push(`Sentença longa (${n} palavras)`);
  if (subs >= 1) factors.push(`${subs} oração(ões) subordinada(s)`);

  return {
    level,
    index,
    factors,
    metrics: { words: n, vocab: +vocab.toFixed(1) },
  };
}
